lib/src/rules/close_sinks.dart

```diff
@@ -0,0 +1,137 @@
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.close_sinks;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:linter/src/linter.dart';
+import 'package:linter/src/utilities/dart_type_utilities.dart';
+
+typedef void _VisitVariableDeclaration(VariableDeclaration node);
+
+const _desc = r'Close instances of `dart.core.Sink`.';
+
+const _details = r'''
+
+**DO** Invoke `close` on instances of `dart.core.Sink` to avoid memory leaks and
+unexpected behaviors.
+
+**BAD:**
+```
+class A {
+  IOSink _sinkA;
+  void init(filename) {
+    _sinkA = new File(filename).openWrite(); // LINT
+  }
+}
+```
+
+**BAD:**
+```
+void someFunction() {
+  IOSink _sinkF; // LINT
+}
+```
+
+**GOOD:**
+```
+class B {
+  IOSink _sinkB;
+  void init(filename) {
+    _sinkB = new File(filename).openWrite(); // OK
+  }
+
+  void dispose(filename) {
+    _sinkB.close();
+  }
+}
+```
+
+**GOOD:**
+```
+void someFunctionOK() {
+  IOSink _sinkFOK; // OK
+  _sinkFOK.close();
+}
+```
+''';
+
+class CloseSinks extends LintRule {
+  CloseSinks() : super(
+      name: 'close_sinks',
+      description: _desc,
+      details: _details,
+      group: Group.errors,
+      maturity: Maturity.experimental);
+
+  @override
+  AstVisitor getVisitor() => new Visitor(this);
+}
+
+class Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  Visitor(this.rule);
+
+  @override
+  void visitVariableDeclarationStatement(VariableDeclarationStatement node) {
+    FunctionDeclaration function =
+        node.getAncestor((a) => a is FunctionDeclaration);
+    node.variables.variables.forEach(_buildVisitVariableCallback(function));
+  }
+
+  @override
+  void visitFieldDeclaration(FieldDeclaration node) {
+    ClassDeclaration classDecl =
+    node.getAncestor((a) => a is ClassDeclaration);
+    node.fields.variables.forEach(_buildVisitVariableCallback(classDecl));
+  }
+
+  _VisitVariableDeclaration _buildVisitVariableCallback(AstNode container) =>
+          (VariableDeclaration sink) {
+        if (!_implementsDartCoreSink(sink.element.type)) {
+          return;
+        }
+
+        List<AstNode> containerNodes = _traverseNodesInDFS(container);
+        Iterable<AstNode> closeSinkInvocations = _findNodesClosingSink(
+            containerNodes, sink);
+
+        if (closeSinkInvocations.isEmpty) {
+          rule.reportLint(sink);
+        }
+      };
+}
+
+Iterable<AstNode> _findNodesClosingSink(List<AstNode> classNodes,
+    VariableDeclaration variable) => classNodes.where(
+    (n) =>
+        n is MethodInvocation &&
+            n.methodName.name == 'close' &&
+            n.target is SimpleIdentifier &&
+            (n.target as SimpleIdentifier).name == variable.name.name);
```

Comment:
I am already doing it here where will be used by two rules: https://github.com/dart-lang/linter/pull/242/commits/6b6e91bcc43b04a31d3108572424b6610147da7c#diff-7d79a7aa6339ada4363b2f05cdca8a4bR33